sorting/count_inversions.cpp

```diff
@@ -0,0 +1,198 @@
+ *   3. Create a recursive function to divide the array into halves and find the answer by summing 
+ *      the number of inversions is the first half, number of inversion in the second half and 
+ *      the number of inversions by merging the two.
+ *   4. The base case of recursion is when there is only one element in the given half.
+ *   5. Print the answer
+ * 
+ */
+#include <iostream>
+#include <vector>
+
+namespace sorting   {
+namespace inversion {
+
+int mergeSort(int* arr, int* temp, int left, int right);
+int merge(int* arr, int* temp, int left, int mid, int right);
+int countInversion(int* arr, const int size);
+void show(int* arr, const int size);
+
+/**
+ * Funtion countInversion() returns the number of inversion
+ * present in the input array. Inversions are an estimate of
+ * how close or far off the array is to being sorted.
+ * 
+ * Number of inversions in a sorted array is 0.
+ * Number of inversion in an array[1...n] sorted in 
+ * non-ascending order is n(n-1)/2, since each pair of elements
+ * contitute an inversion.
+ * 
+ * @param arr   - array, data member of std::vector<int>, input for counting inversions
+ * @param array_size    - number of elementa in the array
+ * 
+ */
+int countInversion(int* arr, const int size){
+    std::vector<int> temp;
+    temp.reserve(size);
+    temp.assign(size,0);
+    return mergeSort(arr, temp.data(), 0, size-1);
+}
+
+/**
+ * 
+ * The mergeSort() function implements Merge Sort, a
+ * Divide and conquer algorithm, it divides the input
+ * array into two halves and calls itself for each 
+ * sub-array and then calls the merge() function to 
+ * merge the two halves.
+ * 
+ * @param arr   - array to be sorted
+ * @param temp  - merged resultant array
+ * @param left  - lower bound of array
+ * @param right - upper bound of array
+ * 
+ */
+int mergeSort(int* arr, int* temp, int left, int right) {
+    int mid = 0, inv_count = 0;
+    if(right>left)
+    {
+        // midpoint to split the array
+        mid = (right + left) / 2;
+        // Add inversions in left and right sub-arrays
+        inv_count += mergeSort(arr, temp, left, mid);   // left sub-array
+        inv_count += mergeSort(arr, temp, mid+1, right);
+
+        // inversions in the merge step
+        inv_count += merge(arr, temp, left, mid, right);
+    }
+    return inv_count;
+}
+
+/**
+ * Function to merge two sub-arrays. merge() function is called 
+ * from mergeSort() to merge the array after it split for sorting 
+ * by the mergeSort() funtion.
+ * 
+ * In this case the merge fuction will also count and return 
+ * inversions detected when merging the sub arrays.
+ * 
+ * @param arr    input array, data-menber of vector
+ * @param temp   stores the resultant merged array
+ * @param left   lower bound of arr[] and left-sub-array
+ * @param mid    midpoint, upper bound of left sub-array,
+ *               (mid+1) gives the lower bound of right-sub-array
+ * @param right  upper bound of arr[] and right-sub-array
```

Comment:
Missing return statement documentation.
```suggestion
 * @param right  upper bound of arr[] and right-sub-array
 * @returns description
```